crates/pallet-domains/src/staking.rs

```diff
@@ -387,119 +559,365 @@ pub(crate) fn do_deregister_operator<T: Config>(
-                };
+/// Unlocks any withdraws that are ready to be unlocked.
+pub(crate) fn do_unlock_funds<T: Config>(
+    operator_id: OperatorId,
+    nominator_id: NominatorId<T>,
+) -> Result<BalanceOf<T>, Error> {
+    Withdrawals::<T>::try_mutate_exists(operator_id, nominator_id.clone(), |maybe_withdrawals| {
+        let withdrawals = maybe_withdrawals.as_mut().ok_or(Error::MissingWithdrawal)?;
+        let Withdrawal {
+            allowed_since_domain_epoch,
+            shares,
+        } = withdrawals.pop_front().ok_or(Error::MissingWithdrawal)?;
+
+        // if there are no withdrawals, then delete the storage as well
+        if withdrawals.is_empty() {
+            *maybe_withdrawals = None
+        }
 
-                let nominator_share =
-                    Perbill::from_rational(nominator.shares, operator.total_shares);
+        let (domain_id, epoch_idx) = allowed_since_domain_epoch.deconstruct();
 
-                let nominator_staked_amount = nominator_share.mul_floor(total_stake);
+        let epoch_completed_block =
+            DomainEpochCompleteAt::<T>::get(domain_id, epoch_idx).ok_or(Error::EpochNotComplete)?;
 
-                let nominator_remaining_amount = nominator_staked_amount
-                    .checked_sub(&withdraw_amount)
-                    .ok_or(Error::BalanceUnderflow)?;
+        let unlock_block_number = epoch_completed_block
+            .checked_add(&T::StakeWithdrawalLockingPeriod::get())
+            .ok_or(Error::BlockNumberOverflow)?;
 
-                if operator_owner == nominator_id {
-                    // for operator owner, the remaining amount should not be less than MinimumOperatorStake,
-                    if nominator_remaining_amount < T::MinOperatorStake::get() {
-                        return Err(Error::MinimumOperatorStake);
-                    }
+        // this is the oldest unconfirmed domain block
+        // we want to release funds only after the unlock period is confirmed.
+        let oldest_domain_block = Pallet::<T>::oldest_receipt_number(domain_id);
+        ensure!(
+            unlock_block_number < oldest_domain_block,
+            Error::UnlockPeriodNotComplete
+        );
 
-                    PendingWithdrawals::<T>::insert(operator_id, nominator_id, withdraw);
+        let staked_hold_id = T::HoldIdentifier::staking_staked(operator_id);
+        let locked_amount = T::Currency::balance_on_hold(&staked_hold_id, &nominator_id);
+        let amount_to_unlock: BalanceOf<T> = {
+            let epoch_share_price =
+                OperatorEpochSharePrice::<T>::get(operator_id, allowed_since_domain_epoch)
+                    .ok_or(Error::MissingOperatorEpochSharePrice)?;
+
+            // if the share price is one, just convert shares to ssc
+            let amount_to_unlock = if epoch_share_price.is_one() {
+                shares.into()
+            } else {
+                epoch_share_price.saturating_reciprocal_mul_floor(shares.into())
```

Comment:
Shouldn't you be dividing here? iiuc `epoch_share_price` is measured in shares/SSC, then you should be dividing `shares/epoch_share_price` to get SSC
On the contrary [here](https://github.com/subspace/subspace/pull/2390/files#diff-bc31bd26a15ac3e6de45a5242e5fc7883136b699fa411a472d0d3eb45c50f0f9R352) above in deposits you multiply `amount*epoch_share_price=SSC*shares/SSC=shares`

Reply:
This is doing what you are expecting it to do. `saturating_reciprocal_mul_floor` is essentially converting `Shares/SSC` to `SSC/Shares` and then multiplying nominator shares to get the SSC staked for the nominator.

For deposits, we just multiply since deposited `SSC` with `Shares/SSC` to get the shares for the nominator

Reply:
Ok, but for deposits you also have the `saturating_reciprocal_mul_floor` so it's correct here but incorrect in deposits then?